compiler/runtime.go

```diff
@@ -24,138 +41,137 @@ type Evaluator interface {
 	EvalFunction(scope Scope) values.Function
 }
 
-type Func interface {
-	Type() semantic.Type
-	EvalString(scope Scope) (string, error)
-	Eval(scope Scope) (values.Value, error)
-	EvalInt(scope Scope) (int64, error)
-	EvalUInt(scope Scope) (uint64, error)
-	EvalFloat(scope Scope) (float64, error)
-	EvalBool(scope Scope) (bool, error)
-	EvalTime(scope Scope) (values.Time, error)
-	EvalDuration(scope Scope) (values.Duration, error)
-	EvalRegexp(scope Scope) (*regexp.Regexp, error)
-	EvalArray(scope Scope) (values.Array, error)
-	EvalObject(scope Scope) (values.Object, error)
-	EvalFunction(scope Scope) (values.Function, error)
-}
-
 type compiledFn struct {
-	root    Evaluator
-	inTypes map[string]semantic.Type
+	root       Evaluator
+	fnType     semantic.Type
+	inputScope Scope
 }
 
-func (c compiledFn) validate(scope Scope) error {
-	// Validate scope
-	for k, t := range c.inTypes {
-		if scope.Type(k) != t {
-			return fmt.Errorf("missing or incorrectly typed value found in scope for name %q", k)
+func (c compiledFn) validate(input values.Object) error {
+	sig := c.fnType.FunctionSignature()
+	properties := input.Type().Properties()
+	if len(properties) != len(sig.Parameters) {
+		return errors.New("mismatched parameters and properties")
+	}
+	for k, v := range sig.Parameters {
+		if properties[k] != v {
+			return fmt.Errorf("parameter %q has the wrong type, expected %v got %v.", k, v, properties[k])
 		}
 	}
 	return nil
 }
 
+func (c compiledFn) buildScope(input values.Object) error {
+	if err := c.validate(input); err != nil {
+		return err
+	}
+	input.Range(func(k string, v values.Value) {
+		c.inputScope[k] = v
+	})
+	return nil
+}
+
 func (c compiledFn) Type() semantic.Type {
-	return c.root.Type()
+	return c.fnType.FunctionSignature().Return
 }
 
-func (c compiledFn) Eval(scope Scope) (values.Value, error) {
-	if err := c.validate(scope); err != nil {
+func (c compiledFn) Eval(input values.Object) (values.Value, error) {
+	if err := c.buildScope(input); err != nil {
 		return nil, err
 	}
 	switch c.Type().Kind() {
 	case semantic.String:
-		return values.NewString(c.root.EvalString(scope)), nil
+		return values.NewString(c.root.EvalString(c.inputScope)), nil
```

Comment:
This code is the definition of `Eval` I am not sure which `Eval` you are referring to.

Reply:
I mean that it should just call `c.root.Eval` instead of doing a type switch. They appear to do the same thing and reducing the number of type switches seems good to me.